mountpoint-s3/src/prefetch/task.rs

```diff
@@ -1,65 +1,82 @@
-    task_handle: Option<RemoteHandle<()>>,
+    _task_handle: RemoteHandle<()>,
     remaining: usize,
-    start_offset: u64,
-    total_size: usize,
+    range: RequestRange,
     part_queue: PartQueue<E>,
+    backpressure_controller: BackpressureController,
 }
 
 impl<E: std::error::Error + Send + Sync> RequestTask<E> {
-    pub fn from_handle(task_handle: RemoteHandle<()>, size: usize, offset: u64, part_queue: PartQueue<E>) -> Self {
+    pub fn from_handle(
+        task_handle: RemoteHandle<()>,
+        range: RequestRange,
+        part_queue: PartQueue<E>,
+        backpressure_controller: BackpressureController,
+    ) -> Self {
         Self {
-            task_handle: Some(task_handle),
-            remaining: size,
-            start_offset: offset,
-            total_size: size,
+            _task_handle: task_handle,
+            remaining: range.len(),
+            range,
             part_queue,
+            backpressure_controller,
         }
     }
 
-    pub fn from_parts(parts: impl IntoIterator<Item = Part>, offset: u64) -> Self {
-        let mut size = 0;
-        let (part_queue, part_queue_producer) = unbounded_part_queue();
-        for part in parts {
-            size += part.len();
-            part_queue_producer.push(Ok(part));
-        }
-        Self {
-            task_handle: None,
-            remaining: size,
-            start_offset: offset,
-            total_size: size,
-            part_queue,
+    // Push a given list of parts in front of the part queue
+    pub async fn push_front(&mut self, parts: Vec<Part>) -> Result<(), PrefetchReadError<E>> {
+        // Merge all parts into one single part by pushing them to the front of the part queue.
+        // This could result in a really big part, but we normally use this only for backward seek
+        // so its size should not be bigger than the prefetcher's `max_backward_seek_distance`.
+        for part in parts.into_iter().rev() {
+            self.remaining += part.len();
+            self.part_queue.push_front(part).await?;
         }
+        Ok(())
     }
 
     pub async fn read(&mut self, length: usize) -> Result<Part, PrefetchReadError<E>> {
         let part = self.part_queue.read(length).await?;
         debug_assert!(part.len() <= self.remaining);
         self.remaining -= part.len();
+
+        // We read some data out of the part queue so the read window should be moved
+        self.backpressure_controller.send_feedback(DataRead(part.len())).await?;
+
+        let next_offset = part.offset() + part.len() as u64;
+        let remaining_in_queue = self.available_offset().saturating_sub(next_offset) as usize;
+        // If the part queue is empty it means we are reading faster than the task could prefetch,
+        // so we should use larger window for the task.
+        if remaining_in_queue == 0 {
+            self.backpressure_controller.send_feedback(PartQueueStall).await?;
```

Comment:
Sending `PartQueueStall` here is not as precise as we may be. Reading from the part queue on line 52 may get blocked, but we would not signal about this event to the controller.

I'd expose `is_empty` method in the PartQueue:
```
    pub fn is_empty(&self) -> bool {
        self.receiver.is_empty()
    }
```

Then call it before a read attempt and send `PartQueueStall` if it returns `true`. Moreover I would immediatelly signal about that to the limiter `self.read_window_updater.send(..)`.  This will make the system more robust as it guarantees that when there is no data in the `PartQueue`, limiter is **always** notified about that.